File: platforms/android/foundation/foundation/src/main/java/io/nativeblocks/foundation/modifier/Border.kt
package io.nativeblocks.foundation.modifier

import androidx.compose.foundation.border
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.nativeblocks.foundation.util.SHAPE_RECTANGLE
import io.nativeblocks.foundation.util.cornerShape
import io.nativeblocks.compiler.type.Modifier
import io.nativeblocks.compiler.type.ModifierData
import androidx.compose.ui.Modifier as UiModifier

@Modifier(
    keyType = "nativeblocks/border",
    name = "Border",
    description = "Draws a border around the block.",
    version = 1,
    versionName = "1",
)
internal fun border(
    @ModifierData(description = "Border width in DP.", defaultValue = "1.0")
    width: Dp = 1.dp,
    @ModifierData(
        description = "Border color in hexadecimal format.",
        defaultValue = "#00000000",
    ) color: Color = Color.Transparent,
    @ModifierData(description = "Shape of the border (rectangle, circle).", defaultValue = "rectangle")
    shape: String = SHAPE_RECTANGLE,
    @ModifierData(description = "Top-start corner radius in DP.", defaultValue = "0.0")
    radiusTopStart: Dp = 0.dp,
    @ModifierData(description = "Top-end corner radius in DP.", defaultValue = "0.0")
    radiusTopEnd: Dp = 0.dp,
    @ModifierData(description = "Bottom-start corner radius in DP.", defaultValue = "0.0")
    radiusBottomStart: Dp = 0.dp,
    @ModifierData(description = "Bottom-end corner radius in DP.", defaultValue = "0.0")
    radiusBottomEnd: Dp = 0.dp,
): UiModifier = UiModifier.border(
    width = width,
    color = color,
    shape = cornerShape(shape, radiusTopStart, radiusTopEnd, radiusBottomStart, radiusBottomEnd),
)

File: platforms/ios/foundation/Sources/NativeblocksFoundation/modifier/Border.swift
import NativeblocksCompiler
import NativeblocksRuntime
import SwiftUI

@Modifier(
    name: "Border",
    keyType: "nativeblocks/border",
    description: "Draws a border around the block.",
    version: 1,
    versionName: "1"
)
struct Border: ViewModifier {

    @ModifierData(description: "Border width.", defaultValue: "1.0")
    var width: CGFloat = 1

    @ModifierData(description: "Border color in hexadecimal format.", defaultValue: "#00000000")
    var color: Color = Color.black.opacity(0)

    @ModifierData(description: "Shape of the border (rectangle, circle).", defaultValue: "rectangle")
    var shape: String = SHAPE_RECTANGLE

    @ModifierData(description: "Top-leading corner radius.", defaultValue: "0.0")
    var radiusTopLeading: CGFloat = 0

    @ModifierData(description: "Top-trailing corner radius.", defaultValue: "0.0")
    var radiusTopTrailing: CGFloat = 0

    @ModifierData(description: "Bottom-leading corner radius.", defaultValue: "0.0")
    var radiusBottomLeading: CGFloat = 0

    @ModifierData(description: "Bottom-trailing corner radius.", defaultValue: "0.0")
    var radiusBottomTrailing: CGFloat = 0

    @Environment(\.layoutDirection) var layoutDirection

    func body(content: Content) -> some View {
        let outline = CornerShape(
            shape: shape,
            topLeading: radiusTopLeading,
            topTrailing: radiusTopTrailing,
            bottomLeading: radiusBottomLeading,
            bottomTrailing: radiusBottomTrailing,
            layoutDirection: layoutDirection
        )
        // a stroke straddles the path, so draw it double width and clip the outer half away
        return content.overlay(outline.stroke(color, lineWidth: width * 2).clipShape(outline))
    }
}
